### PR TITLE
Add Router Validations

### LINKED ISSUE #152: Verify not Unreachable Selectors In Router AST

### LINKED ISSUE #153: Verify No Duplicate Selectors In Router AST

### PR DESCRIPTION
**Pull request by leomassazza**
Fix #152
Fix #153

### CODE REVIEW

File: packages/deployer/internal/ast/ast-helper.js [hunk 1]
```diff
@@ -100,11 +104,21 @@ function getCaseSelectors(contractName, ast) {
...
+
   const items = [];
-  for (const assignment of findAll('YulCase', contractNode)) {
-    if (assignment.value === 'default') continue;
-    if (assignment.value.value === '0') continue;
-    items.push(assignment.value.value);
+  for (const caseSelector of findAll('YulCase', contractNode)) {
+    if (caseSelector.value === 'default') continue;
+    if (caseSelector.value.value === '0') continue;
+    const caseAssignment = findAll('YulAssignment', caseSelector);
```
Comment (leomassazza) [inline packages/deployer/internal/ast/ast-helper.js:113]: adds more data to the selector (Pointed constant name and value)

File: packages/deployer/internal/ast/router-ast-validator.js [hunk 2]
```diff
@@ -30,14 +30,43 @@ async function findMissingSelectorsInAST(contracts) {
...
+async function findUnreachableSelectorsInAST(contracts, modules) {
+  const routerSelectors = getCaseSelectors('Router', contracts['Router']);
+  const moduleAddresses = [];
+  for (const [moduleName, moduleData] of Object.entries(modules)) {
+    moduleAddresses[toPrivateConstantCase(moduleName)] = moduleData.deployedAddress;
+  }
+
+  const errors = [];
+  routerSelectors.forEach((s) => {
+    if (!moduleAddresses[s.value.name] || moduleAddresses[s.value.name] !== s.value.value.value) {
```
Comment (leomassazza) [inline packages/deployer/internal/ast/router-ast-validator.js:42]: Checks that the address of the contract pointed is the one of the contract.

File: packages/deployer/internal/router-helper.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+/**
```
Comment (leomassazza) [inline packages/deployer/internal/router-helper.js:1]: Moved to a utils to reuse in AST validation

File: packages/deployer/internal/ast/router-ast-validator.js (hunk 2 above)
Comment (ajsantander) [reply packages/deployer/internal/ast/router-ast-validator.js:42]: Could this also check that the hardcoded module name is correct?

Right now of you change for example:
`case 0xa40674b7 { result := _SOME_MODULE } // SomeModule.getSomeValue()`
to
`case 0xa40674b7 { result := _OWNER_MODULE } // SomeModule.getSomeValue()`

It does not fail.

**Review by ajsantander — APPROVED** [review]

File: packages/deployer/internal/ast/router-ast-validator.js (hunk 2 above)
Comment (leomassazza) [reply packages/deployer/internal/ast/router-ast-validator.js:42]: good catch. It doesn't seem that trivial, I might create another issue for that.

Comment (leomassazza) [reply packages/deployer/internal/ast/router-ast-validator.js]: done

File: packages/deployer/internal/ast/router-ast-validator.js [hunk 4]
```diff
@@ -30,14 +30,61 @@ async function findMissingSelectorsInAST(contracts) {
...
+    if (
+      !moduleAddresses[s.value.name] ||
+      moduleAddresses[s.value.name].address !== s.value.value.value
+    ) {
+      errors.push({
+        msg: `Selector ${s.selector} not reachable. ${s.value.name} pointing to ${
+          s.value.value.value
+        } instead of ${moduleAddresses[s.value.name]}`,
+      });
+    } else {
```
Comment (leomassazza) [inline packages/deployer/internal/ast/router-ast-validator.js:54]: This section will check if the pointed contract has a function with that selector

**Review by mjlescano — APPROVED** [review]

**Pull request merged**